### PR TITLE
Add auto-rollback with per-stack health checks

### LINKED ISSUE #6: Auto-rollback to last known good state on deploy failure
## Summary

Implement ArgoCD/Flux-style auto-rollback: track the last successful commit per stack, health-check after deploy, and automatically rollback to the last known good version on failure.

## Motivation

Currently, if a deploy produces unhealthy containers, the error is logged and metrics are recorded, but the broken state persists until someone pushes a fix or reverts in git. ArgoCD and Flux (for Helm releases) handle this automatically by rolling back to the last successful revision without requiring a new git commit.

### 1. Per-stack deployment history

Extend the state file (currently just a single commit hash) to track `lastSuccessfulCommit` per stack:

```json
{
  "last_hash": "abc123",
  "stacks": {
    "traefik": {"last_successful_commit": "abc123"},
    "vault": {"last_successful_commit": "abc123"}
  }
}
```

### 2. Post-deploy health check

After `docker compose up -d`, wait for containers to reach healthy state:

- Run `docker compose ps --format json` to check container states
- Configurable timeout per stack (default 60s, overridable via `health_check_timeout` in gitops.yaml)
- Healthy = all containers in `running` state (or `healthy` if they define a HEALTHCHECK)

### 3. Auto-rollback flow

```
deploy stack at HEAD commit
  → health check passes → update lastSuccessfulCommit
  → health check fails →
      1. git show <lastSuccessfulCommit>:<path>/docker-compose.yml > tempdir
      2. docker compose -f tempdir/docker-compose.yml up -d
      3. mark stack as degraded (dockcd_stack_healthy=0)
      4. log: "stack X rolled back from <bad_commit> to <good_commit>"
      5. on next NEW commit touching this stack → try again
```

### 4. Config additions

```yaml
hosts:
  server04:
    stacks:
      - name: traefik
        path: server04/traefik
        health_check_timeout: 60s   # default 60s
        auto_rollback: true          # default true
```

### 5. Edge cases

- **First deploy (no history)**: no rollback possible — just mark as failed
- **Rollback also fails**: mark as degraded, don't retry until next git commit
- **Git working tree state**: use `git show` to extract files to a temp dir rather than modifying the working tree, so the repo stays clean at HEAD

## Out of scope (for now)

- Rollback notifications (Slack, webhook) — can use Grafana alerts on `dockcd_stack_healthy=0`
- Manual rollback trigger via `/force-deploy` endpoint
- Rollback history / audit log beyond structured logs

## References

- [ArgoCD Automated Sync + Self Heal](https://argo-cd.readthedocs.io/en/stable/user-guide/auto_sync/)
- [Flux HelmRelease Remediation](https://fluxcd.io/flux/components/helm/helmreleases/#configuring-failure-handling)

### PR DESCRIPTION
**Pull request by Mohitsharma44**
## Summary
- **Health checks**: After `docker compose up -d`, polls `docker compose ps --format json` to verify all containers are running/healthy. Configurable timeout per stack (default 60s).
- **Auto-rollback**: On health check failure, extracts the last known good compose files via `git show` to a temp dir and redeploys from there. Stack is marked degraded (`dockcd_stack_healthy=0`).
- **Per-stack state**: State file migrated from plain text to JSON with backward compatibility. Tracks `lastSuccessfulCommit` per stack.
- **Config additions**: `health_check_timeout` and `auto_rollback` per stack in gitops.yaml
- **Edge cases**: No history = no rollback (just mark failed). Rollback failure = mark degraded, retry on next commit.

Closes #6

## Test plan
- [x] `go test ./...` — all unit tests pass (including 6 new rollback tests)
- [x] `go test -tags integration ./internal/reconciler/` — integration tests pass
- [x] `go vet ./...` — clean
- [x] `go build ./...` — builds successfully
- [ ] CI passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/reconciler/reconciler.go [hunk 1]
```diff
@@ -207,18 +230,131 @@ func (r *Reconciler) reconcile(ctx context.Context) error {
...
+// Only hard failures (deploy error, failed rollback) halt the pipeline.
+func (r *Reconciler) deployGroups(ctx context.Context, groups [][]deploy.Stack, commitHash string) error {
+	var rolledBack bool
+	for i, group := range groups {
+		g, gCtx := errgroup.WithContext(ctx)
+		for _, stack := range group {
+			g.Go(func() error {
+				return r.deployStack(gCtx, stack, commitHash)
+			})
+		}
```
[the remark this answers is not in the document]
Comment (Mohitsharma44) [reply internal/reconciler/reconciler.go]: Fixed. Replaced `errgroup.WithContext` with a plain `errgroup.Group` so that `ErrRolledBack` from one stack no longer cancels the context for sibling stacks in the same parallel group.

File: internal/reconciler/reconciler.go [hunk 2]
```diff
@@ -207,18 +230,133 @@ func (r *Reconciler) reconcile(ctx context.Context) error {
...
+		for _, stack := range group {
+			g.Go(func() error {
+				return r.deployStack(ctx, stack, commitHash)
+			})
+		}
+		if err := g.Wait(); err != nil {
+			if errors.Is(err, ErrRolledBack) {
+				rolledBack = true
+				continue // stack is running (rolled back), proceed with downstream
+			}
```
[the remark this answers is not in the document]
Comment (Mohitsharma44) [reply internal/reconciler/reconciler.go]: Fixed. ErrRolledBack is now handled inside the goroutine via an atomic.Bool flag, so it no longer masks real deploy failures in g.Wait().

File: internal/reconciler/reconciler.go [hunk 3]
```diff
@@ -207,18 +230,133 @@ func (r *Reconciler) reconcile(ctx context.Context) error {
...
+	// Health check (skip if no OutputRunner or timeout is 0).
+	if r.outputRunner != nil && stack.HealthCheckTimeout > 0 {
+		dir := filepath.Join(r.repoDir, stack.Path)
+		if err := deploy.HealthCheck(ctx, dir, stack.Name, stack.HealthCheckTimeout, r.outputRunner); err != nil {
+			r.logger.Warn("health check failed",
+				"stack", stack.Name, "error", err)
+
+			if !stack.AutoRollback {
+				r.markStackHealth(ctx, stack.Name, false)
+				return fmt.Errorf("stack %q unhealthy, auto_rollback disabled", stack.Name)
```
[the remark this answers is not in the document]
Comment (Mohitsharma44) [reply internal/reconciler/reconciler.go]: Fixed. The error now wraps the health check error: `fmt.Errorf("stack %q unhealthy, auto_rollback disabled: %w", stack.Name, err)`.

File: internal/git/poller.go [hunk 4]
```diff
@@ -72,7 +99,75 @@ func (p *Poller) LoadState() error {
...
+			return fmt.Errorf("extracting %s:%s: %w", commit, file, err)
+		}
+
+		// Write to destDir, preserving the relative path under pathPrefix.
+		relPath, err := filepath.Rel(pathPrefix, file)
+		if err != nil {
+			return fmt.Errorf("computing relative path for %s: %w", file, err)
+		}
+		destPath := filepath.Join(destDir, relPath)
+		if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
```
[the remark this answers is not in the document]
Comment (Mohitsharma44) [reply internal/git/poller.go:164]: Acknowledged. This tool targets Linux-only environments (Docker Compose on Linux hosts), so filepath and git paths are both slash-separated. Adding a Windows-safe path layer would be dead code for now, but noted for future portability.

File: internal/deploy/health.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		if len(bytes.TrimSpace(line)) == 0 {
+			continue
+		}
+		var c ContainerStatus
+		if err := json.Unmarshal(line, &c); err != nil {
+			return nil, fmt.Errorf("parsing container status: %w", err)
+		}
+		containers = append(containers, c)
+	}
+	return containers, scanner.Err()
```
[the remark this answers is not in the document]
Comment (Mohitsharma44) [reply internal/deploy/health.go:110]: In practice, `docker compose ps --format json` output per container is well under 1KB (name, state, health, a few labels). The 64KB default scanner buffer is more than sufficient. Switching to json.Decoder would add complexity for a scenario that won't occur.

File: internal/reconciler/reconciler_test.go [hunk 6]
```diff
@@ -431,7 +479,7 @@ func TestDeployOrderRespectsDependencies(t *testing.T) {
 	var order []string
 	runner := func(ctx context.Context, dir, name string, args ...string) error {
 		// Only track "compose up" calls, not "compose pull".
-		if len(args) > 0 && args[0] == "compose" && len(args) > 1 && args[1] == "up" {
+		if args[0] == "compose" && slices.Contains(args, "up") {
```
[the remark this answers is not in the document]
Comment (Mohitsharma44) [reply internal/reconciler/reconciler_test.go]: Fixed. Added `len(args) > 0` guard before indexing `args[0]` in both unit and integration test runners.
